src/model.rs

```diff
@@ -787,88 +792,79 @@ impl Collections {
     ///  VJ:2 (Mon-Fri)                                       12:00-13:00       14:00-15:00
     ///  VJ:3 (Sat-Sun)                                       12:30-13:30       14:30-15:30
     ///
-    /// Example 4 might be a valid use case of stay-in but is not handled in the
-    /// current implementation since the validity patterns are different; this
-    /// is undefined behavior (most likely, some stay-in will be forbidden by
-    /// the default rule).
+    /// Example 4 is a valid use case of stay-in
+    /// The pickup/dropoff will be possible between VJ:1 and VJ:2/VJ:3
     pub fn enhance_pickup_dropoff(&mut self) {
-        let mut vj_idxs: Vec<Idx<VehicleJourney>> = self
-            .vehicle_journeys
+        let mut allowed_last_pick_up_vj = HashSet::<_>::new();
+        let mut allowed_first_drop_off_vj = HashSet::<_>::new();
+
+        let can_chain_without_overlap = |prev_vj: &VehicleJourney, next_vj: &VehicleJourney| {
+            let last_stop = &prev_vj.stop_times.last();
+            let first_stop = &next_vj.stop_times.first();
+            match (last_stop, first_stop) {
+                (Some(last_stop), Some(first_stop)) => {
+                    if last_stop.stop_point_idx == first_stop.stop_point_idx {
+                        // We can discard when the stop points are identicals (see
+                        // Example 1 above).
+                        false
+                    } else {
+                        let prev_cal = self.calendars.get(&prev_vj.service_id);
+                        let next_cal = self.calendars.get(&next_vj.service_id);
+                        // for the stay in to be possible the vj should have at least one date in common
+                        prev_cal.map_or(false, |prev_cal| next_cal.map_or(false, |next_cal|prev_cal.overlaps(&next_cal))) &&
+                        // The stay in is not really possible when timing overlaps
+                        // between arrival of first vehicle journey and departure of
+                        // next vehicle journey (see Example 2 above).
+                        last_stop.departure_time <= first_stop.arrival_time
```

Comment:
If one is past-midnight and the next is at the beginning of the day after, I guess it won't work but it "only" matters for night-bus lines
:see_no_evil: becomes a bit too complex for the gain I think.

Reply:
after giving it some thoughts, past midnight are usually implemented as time > 24h (like `"25:00:00"` for 1am) no? and this case is handled no :thinking: ?

Reply:
It would not be handled I think, as the first VJ would end at `25:00` and the second would start at `02:00`.
They would be possible stay-in, but we wouldn't "see" that.
But that's OK for now (and possibly for ever :stuck_out_tongue_closed_eyes: )
I do not resolve the comment just to have it quick if we ever come back later to that work to improve those tricky cases.